### ISSUE TITLE
Docker Windows proxy not working

### ISSUE THREAD

**Issue by kzhui125** [docker/for-win#6260]
- Diagnostics ID: 88B3E0B0-A0FB-431B-83CA-3502BCCBD473/20200407094643
  - Windows Version: Windows Server 2019 datacenter
  - Docker Desktop Version: 2.2.0.5

I have set http proxy in settings, but docker pull don't use my proxy.

When I use Docker for mac, it works fine.
This is a problem of Docker for Windows.

<img width="799" alt="3" src="https://user-images.githubusercontent.com/11925432/78656597-371c9a80-78fa-11ea-89d6-b08c8a30cbf8.png">

**Comment by mat007** [comment]
Thanks for the report, is this the same as #5606 which has been fixed on the Edge builds?

**Comment by kzhui125** [comment]
@mat007 https://github.com/docker/for-win/issues/5606 says 2.2.0.3 fix that issue, I am using 2.2.0.5

And C:\Users\Administrator\AppData\Roaming\Docker\http_proxy.json is valid in my case.

**Comment by mat007** [comment]
This definitely rings a bell as something we’ve fixed recently.
I’ll try and find a bit of time later today to check.

**Comment by mat007** [comment]
A quick test with the latest Edge shows it’s at least trying to use the proxy I’ve set in the settings (I don’t have a proxy running):
```
$ docker pull golang:1.14.1
Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp 127.0.0.1:3129: connect: connection refused
```
whereas with Stable 2.2.0.5 I get a different error
```
$ docker pull golang:1.14.1
Error response from daemon: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```
but it’s still trying to go through I believe.

What’s the output you’re seeing?

**Comment by kzhui125** [comment]
It seems that when you set proxy other than 127.0.0.1 , it works.
If you set proxy 127.0.0.1, it doesn't work.

You can create a test http server, go build and deploy.

```go
package main

import (
	"log"
	"net/http"

	"gopkg.in/elazarl/goproxy.v1"
)

func main() {
	proxy := goproxy.NewProxyHttpServer()
	proxy.Verbose = true
	log.Fatal(http.ListenAndServe(":1081", proxy))
}
```

If I deploy the http proxy to another machine (192.168.x.x), it works.
If I deploy the http proxy to localhost, I get this error.

Error response from daemon: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

**Comment by mat007** [comment]
Sorry it took so long to get back to you.

There seem to indeed be an issue with `127.0.0.1` (or `localhost`) as we used to have a special case for it in order to redirect to the host (the `docker pull` is happening in the VM).
We have added a ticket in our backlog to investigate why it’s no longer there and put it back.

In the meanwhile a workaround would be to use `host.docker.internal` in the proxy configuration instead of `127.0.0.1`. Does this work?

**Comment by mat007** [comment]
Actually I’ve just realized this is a duplicate of #5715.
I’m going to close this issue in favor of that one if you don’t mind?

**Comment by kzhui125** [comment]
@mat007 host.docker.internal works, thanks.

**Comment by djs55** [comment]
The fix for this has been included in [2.4.0.0 stable](https://desktop.docker.com/win/stable/48506/Docker%20Desktop%20Installer.exe).
